sdk/eventhub/Azure.Messaging.EventHubs/design/api-skeleton-january-2020.md

```diff
@@ -0,0 +1,530 @@
+# .NET Event Hubs Client: Skeleton (January, 2020)
+
+This skeleton represents the public API surface of the .NET Event Hubs client libraries in skeletal form, as they were proposed and reviewed for the preview 6 and January 2020 milestone releases.
+
+## Package: Azure.Messaging.EventHubs
+
+### `Azure.Messaging.EventHubs`
+```csharp
+public class EventHubConnection : IAsyncDisposable 
+{
+    public bool IsClosed { get; protected set; }
```

Comment:
Should public properties like this be marked virtual so that they can be mocked? I realize the guidelines only mention public methods, but I can imagine some scenarios where a consumer of the class may want to test scenarios that require mocking these properties as well.

Reply:
The intent is for this to be mocked via use of the protected setter.  As a general rule, I try to avoid virtual properties unless there is a strong use case, which I find tends to favor loosening visibility on the setter.  While it's a bit less convenient to have to create a stub to mock the property, I think the trade-off of giving the compiler more room to optimize since property access is intended to be a light weight operation.

In this particular case, `IsClosed` is intended to be a computed property that isn't likely to be interesting for mocking outside of a few specific scenarios.

Reply:
Just out of curiosity, I ran some benchmarks and I confirmed that the perf difference is significant, relatively speaking:

```
|        Method |      Mean |     Error |    StdDev |    Median | Ratio |
|-------------- |----------:|----------:|----------:|----------:|------:|
|    VirtualGet | 0.4961 ns | 0.0128 ns | 0.0113 ns | 0.4940 ns | 1.000 |
| NonVirtualGet | 0.0008 ns | 0.0020 ns | 0.0017 ns | 0.0000 ns | 0.002 |
```

The JIT Asm difference, which shows the get call inlined for the non-virtual case, can be found [here](https://sharplab.io/#gist:36c59edc408a8b85d2ce489e4a9763f1)

Reply:
I've only got a shallow understanding of most of the optimizations in the modern compiler, but that confirms what I'd expect.  My understanding is that the compiler can be far more aggressive about inlining simple property access when there's no VTable indirection, especially in the case where setting is infrequent.

Thanks for pushing on that.   Its enough black magic that is constantly changing, so the potential habits and guidance from versions past can absolutely be no longer relevant and totally overlooked.